typescript-for-haskellers/article.md

```diff
@@ -0,0 +1,1002 @@
+const a: NoModifiersA = { x: 1, y: 2 }; // Ok
+a.x = 2; // Ok
+```
+
+Let's present some complex examples of mapped types usage.
+
+Defining type which contains string and numerical properties.
+
+```typescript
+type Point = {
+  x: number;
+  y: number;
+  0: string;
+}
+```
+
+Now, using mapped types we will create a type which will pick only some properties.
+Word `extends` here say that generic `K` will contain a subset of union properties keys `T`.
+And the resulting type will contain only properties from `K`.
+
+```typescript
+type Pick<T, K extends keyof T> = {
+  [P in K]: T[P];
+};
+
+type PointValues = Pick<Point, "x" | 0> // type PointValues = { x: number; 0: number }
+```
+
+The next example will be more complex.
+We will try to create the type which produce type with getters for all string properties.
+
+To implement it lets firstly define helper type `Getter`.
+Capitalize is a build in type which may be applied for the string type.
+So, for `T` which extends string we can produce getter specific name.
+
+```typescript
+type Getter<T extends string> = `get${Capitalize<T>}`;
+```
+
+And finally we can implement `Getters` type.
+Type `Extract` extract from type all union members which are assignable to some type.
+So, here for each property name `K` in `T` which is assignable to `string` we create getter property name with type `() => T[K]`.
+
+```typescript
+type Getters<T> = {
+    [K in keyof T as Getter<Extract<K, string>>]: () => T[K];
+}
+
+type PointGetters = Getters<Point> // type PointValues = { getX: () => number; getY: () => number; }
+```
+
+#### Conditional types
+
+Another great feature is conditional types.
+It allows you to choose the type based on some condition expressed as another type.
+Using `extends` word you are able to make a condition `T extends U ? X : Y`.
+It also allows you to not only choose from two types, but from many types `T extends U ? X : T extends W ? Y : Z`.
+With such functionality you may implement things which may resemble type families although not quite right.
+Anyway in conjunction with mapped types it is a great feature to manage your types.
+One thing to note are union types.
+Conditional types with union types are called Distributive Conditional Types and their work may look a little confused.
+But we will explain them on examples.
+
+Here is the base example of conditional types.
+Type `B` extends `A` but type `C` doesn't.
+And `Condition` type will return one type if `T extends V` and another one otherwise.
+
+```typescript
+type A = { x: number };
+type B = { x: number; y: number };
+type C = { y: number };
+
+type Condition<T, V> = T extends V ? number : string;
+
+type N = Condition<B, A>; // type N = number
+type E = Condition<C, A>; // type E = string
+```
+
+Let's move to the Distributive Conditional Types.
+`Exclude` type may looks strange, but what if we will pass union types here?
```

Comment:
I have to say I cannot understand this.

Reply:
Hm, I just say that `Exclude` is strange but with `union types` as type argument it is not strange and below is an explanation